### PR TITLE
Divide build output into layers to optimize layer cache hits

### PR DESCRIPTION
**Pull request by fg-j**
## Summary

Resolves https://github.com/paketo-buildpacks/dotnet-core/issues/610

## Checklist

* [ ] I have viewed, signed, and submitted the Contributor License Agreement.
* [ ] I have linked issue(s) that this PR should close using keywords or the Github UI (See [docs](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue))
* [ ] I have added an integration test, if necessary.
* [ ] I have reviewed the [styleguide](https://github.com/paketo-buildpacks/community/blob/main/STYLEGUIDE.md) for guidance on my code quality.
* [ ] I'm happy with the commit history on this PR (I have rebased/squashed as needed).

### CODE REVIEW

File: integration/default_apps_test.go [hunk 1]
```diff
@@ -20,6 +21,7 @@ func testDefaultApps(t *testing.T, context spec.G, it spec.S) {
 		pack       occam.Pack
 		docker     occam.Docker
 	)
+	format.MaxLength = 0
```
Comment (ForestEckhardt) [inline integration/default_apps_test.go:24]: You may want to move this into the init_test.go

File: internal/assets_json.go [hunk 2]
```diff
@@ -0,0 +1,98 @@
+package internal
+
+import (
+	"encoding/json"
+	"fmt"
+)
+
+type AssetsJSON struct {
```
Comment (fg-j) [inline internal/assets_json.go:8]: TODO: Rename to ProjectAssetsJSON

File: output_slicer.go [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+	"path/filepath"
+	"strings"
+
+	"github.com/paketo-buildpacks/dotnet-publish/internal"
+	"github.com/paketo-buildpacks/packit/v2"
+)
+
+type namedSlices map[string]map[string]string
+
+func addElement(slices namedSlices, name, element string) namedSlices {
```
Comment (fg-j) [inline output_slicer.go:16]: TODO: Rename - addPath

**Comment by fg-j** [comment]
Something I want to discuss with .NET maintainers:
In the current implementation, if layer slicing fails, the [build fails](https://github.com/paketo-buildpacks/dotnet-publish/blob/b94ce7da28dc233475f2f9e00b6093e9a1122408/build.go#L127).

I wonder whether it's better to warn and proceed with the build, since this is extra optimization that does not have any bearing on the viability of the built app artifact. The schema for the `project.assets.json` isn't documented anywhere I could find, so I fear that the ProjectAssetsJSON parsing is sort of brittle. Another alternative would be to add an environment variable that allows users to turn off output slicing. Something like `BP_DOTNET_DISABLE_BUILD_OUTPUT_SLICING`, which really rolls off the tongue.

What do you think, @paketo-buildpacks/dotnet-core ?

**Comment by fg-j** [comment]
Ping @paketo-buildpacks/dotnet-core-maintainers Now that we've rolled back #305, this is unblocked.

File: integration/output_slicing_test.go [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+				Expect(string(contents)).To(ContainSubstring("My Cool V1 API"))
+				modified.Close()
+
+				image, logs, err = build.Execute(name, source)
+				Expect(err).NotTo(HaveOccurred(), logs.String())
+				images[image.ID] = ""
+				Expect(logs).To(ContainLines(
+					"Reusing 2/3 app layer(s)",
+					"Adding 1/3 app layer(s)",
+				), logs.String())
```
[the remark this answers is not in the document]
Comment (fg-j) [reply integration/output_slicing_test.go:100]: These come from `pack` itself, during the exporter phase.

File: output_slicer.go [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+			file := filepath.Base(string(dep.Runtime))
+			if file != "" && file != "_._" {
+				slices = addPath(slices, dep.Type, file)
+			}
+
+			for _, rt := range dep.RuntimeTargets {
+				slices = addPath(slices, dep.Type, filepath.Base(rt.FileName))
+			}
+		}
+		// TODO: What if same dependency gets filed under two different types?
```
[the remark this answers is not in the document]
Comment (fg-j) [reply output_slicer.go:66]: It seems like the lifecycle is smart enough to deduplicate paths. See the [slices implementation](https://github.com/buildpacks/lifecycle/blob/aaf097aa259bc48e4b35b192492cbc3b8ebb8fd7/layers/slices.go#L162). Given that, I think deduplicating here would add unnecessary complexity.

**Comment by fg-j** [comment]
@sophiewigmore When it comes to env var naming, maybe it's wise to keep it namespaced to `BP_DOTNET` for now -- since this may or may not end up being a cross-cutting Paketo env var.

**Comment by sophiewigmore** [comment]
`BP_DOTNET_DISABLE_BUILDPACK_OUTPUT_SLICING` works for me! :P

**Comment by sophiewigmore** [comment]
`BP_DOTNET_DISABLE_BUILDPACK_OUTPUT_SLICING` works for me!

**Review by sophiewigmore — APPROVED** [review]

**Pull request merged**